### PR TITLE
koordlet: refactor pod/container metrics to tsdb

### LINKED ISSUE #1333: [proposal] refactor pod/container metrics to tsdb
**What is your proposal**:
see https://github.com/koordinator-sh/koordinator/issues/1241 for full description.

**Why is this needed**:
refactor pod/container metircs cache to tsdb for better performance and expansibil

**Is there a suggested solution, if so, please add it**:

### PR DESCRIPTION
**Pull request by LambdaHJ**
### Ⅱ. Does this pull request fix one issue?

fixes #1333

### V. Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `make test`

### CODE REVIEW

File: pkg/koordlet/metricsadvisor/collectors/podresource/pod_resource_collector.go [hunk 1]
```diff
@@ -181,6 +196,7 @@ func (p *podResourceCollector) collectPodResUsed() {
 func (p *podResourceCollector) collectContainerResUsed(meta *statesinformer.PodMeta) {
 	klog.V(6).Infof("start collectContainerResUsed")
 	pod := meta.Pod
+	containerMetrics := make([]metriccache.MetricSample, 0)
```
Comment (saintube) [inline pkg/koordlet/metricsadvisor/collectors/podresource/pod_resource_collector.go:199]: The same as the `podMetrics`. It seems that the variable should be newed inside the loop.

File: pkg/koordlet/statesinformer/states_nodemetric.go [hunk 2]
```diff
@@ -485,25 +491,118 @@ func (r *nodeMetricInformer) collectNodeAggregateMetric(endTime time.Time, aggre
...
+			ResourceList: corev1.ResourceList{
+				corev1.ResourceCPU:    *resource.NewMilliQuantity(int64(cpuUsed*1000), resource.DecimalSI),
+				corev1.ResourceMemory: *resource.NewQuantity(int64(memUsed), resource.BinarySI),
+			},
+		},
+	}
+
+	value, exist := r.metricCache.Get(koordletutil.GPUDeviceType)
+	if !exist {
+		klog.Warningf("get node device info failed, error: %v", err)
```
Comment (zwzhang0107) [inline pkg/koordlet/statesinformer/states_nodemetric.go:537]: this warning should be ignored, node may not have gpu

File: pkg/koordlet/metricsadvisor/devices/gpu/collector_gpu_linux.go [hunk 3]
```diff
@@ -181,7 +179,8 @@ func (g *gpuDeviceManager) getNodeGPUUsage() []metriccache.MetricSample {
 	return gpuMetrics
 }
 
-func (g *gpuDeviceManager) getTotalGPUUsageOfPIDs(pids []uint32) []metriccache.GPUMetric {
```
Comment (zwzhang0107) [inline pkg/koordlet/metricsadvisor/devices/gpu/collector_gpu_linux.go:184]: we can pass container/pod id, and build metric meta inside
`getPodOrContainerGPUUsage(id, pids)`

File: pkg/koordlet/metricsadvisor/collectors/podresource/pod_resource_collector.go [hunk 4]
```diff
@@ -143,45 +143,58 @@ func (p *podResourceCollector) collectPodResUsed() {
 		// do subtraction and division first to avoid overflow
 		cpuUsageValue := float64(currentCPUUsage-lastCPUStat.CPUUsage) / float64(collectTime.Sub(lastCPUStat.Timestamp))
 
-		memUsageValue := memStat.Usage()
+		cpuUsageMetric, err := metriccache.PodCPUUsageMetric.GenerateSample(
+			metriccache.MetricPropertiesFunc.Pod(uid), collectTime, cpuUsageValue)
+		if err != nil {
+			klog.Warningf("generate pod cpu metrics failed, err %v", err)
```
Comment (saintube) [inline pkg/koordlet/metricsadvisor/collectors/podresource/pod_resource_collector.go:149]: `klog.V(4).Infof()` is preferred. How about `klog.V(4).Infof("failed to generate pod cpu metrics for pod %s, err %v", podKey, err)`? Same below.

File: pkg/koordlet/metricsadvisor/collectors/podresource/pod_resource_collector.go [hunk 5]
```diff
@@ -227,33 +240,31 @@ func (p *podResourceCollector) collectContainerResUsed(meta *statesinformer.PodM
...
-			},
-			MemoryUsed: metriccache.MemoryMetric{
-				// 1.0 kB Memory = 1024 B
-				MemoryWithoutCache: *resource.NewQuantity(memUsageValue, resource.BinarySI),
-			},
+		memUsageValue := memStat.Usage()
+		memUsageMetric, memErr := metriccache.ContainerMemUsageMetric.GenerateSample(
+			metriccache.MetricPropertiesFunc.Container(containerStat.ContainerID), collectTime, float64(memUsageValue))
+		if cpuErr != nil || memErr != nil {
+			klog.Warningf("generate container %s cpu metrics failed, cpu: %v, mem:%v", containerKey, cpuErr, memErr)
```
Comment (saintube) [inline pkg/koordlet/metricsadvisor/collectors/podresource/pod_resource_collector.go:250]: `... container %s cpu metrics failed, cpu: %v, mem:%v`
=> `... container %s metrics failed, cpu: %v, mem: %v`

File: pkg/koordlet/resmanager/cpu_burst.go [hunk 6]
```diff
@@ -409,13 +405,17 @@ func (b *CPUBurst) cfsBurstAllowedByLimiter(burstCfg *slov1alpha1.CPUBurstConfig
...
 	containerCPUUsage := containerCPULimit
-	containerRes := b.resmanager.collectContainerResMetricLast(containerID)
-	if containerRes.Error != nil {
-		klog.Warningf("failed to get container %v resource metric, error %v", *containerID, containerRes.Error)
-	} else if containerRes.Metric == nil || containerRes.AggregateInfo == nil {
-		klog.Warningf("container %v resource metric is nil, detail %v", *containerID, containerRes)
+
+	querMeta, err := metriccache.ContainerCPUUsageMetric.BuildQueryMeta(metriccache.MetricPropertiesFunc.Container(*containerID))
+	if err != nil {
+		klog.Warningf("get container metric queryMeta failed, error: %v", err)
```
Comment (saintube) [inline pkg/koordlet/resmanager/cpu_burst.go:411]: please log with container info (e.g. containerID). same below

File: pkg/koordlet/statesinformer/states_nodemetric.go [hunk 7]
```diff
@@ -485,25 +491,115 @@ func (r *nodeMetricInformer) collectNodeAggregateMetric(endTime time.Time, aggre
...
 		return nil
 	}
 	podUID := string(podMeta.Pod.UID)
-	queryResult := r.metricCache.GetPodResourceMetric(&podUID, queryParam)
-	if queryResult.Error != nil {
-		klog.Warningf("get pod %v resource metric failed, error %v", podUID, queryResult.Error)
+
+	queryer, err := r.metricCache.Querier(*queryParam.Start, *queryParam.End)
+	if err != nil {
+		klog.Warningf("get query failed, error %v", err)
```
Comment (saintube) [inline pkg/koordlet/statesinformer/states_nodemetric.go:502]: please log with pod info (e.g. pod uid or namespaced name)

File: pkg/koordlet/statesinformer/states_nodemetric.go [hunk 8]
```diff
@@ -485,25 +491,115 @@ func (r *nodeMetricInformer) collectNodeAggregateMetric(endTime time.Time, aggre
...
-		PodUsage:  *convertPodMetricToResourceMap(queryResult.Metric),
+		PodUsage: slov1alpha1.ResourceMap{
+			ResourceList: corev1.ResourceList{
+				corev1.ResourceCPU:    *resource.NewMilliQuantity(int64(cpuUsed*1000), resource.DecimalSI),
+				corev1.ResourceMemory: *resource.NewQuantity(int64(memUsed), resource.BinarySI),
+			},
+		},
 	}
+
+	value, exist := r.metricCache.Get(koordletutil.GPUDeviceType)
```
Comment (saintube) [inline pkg/koordlet/statesinformer/states_nodemetric.go:535]: This is a node-level resource and there could be many nodes without GPU. We'd better check if the GPU metric collection is necessary outside the `collectPodMetric()` and call `collectPodGPUMetric()` individually.

File: pkg/koordlet/metricsadvisor/collectors/podresource/pod_resource_collector.go [hunk 9]
```diff
@@ -143,45 +143,58 @@ func (p *podResourceCollector) collectPodResUsed() {
...
-			},
-			MemoryUsed: metriccache.MemoryMetric{
-				// 1.0 kB Memory = 1024 B
-				MemoryWithoutCache: *resource.NewQuantity(memUsageValue, resource.BinarySI),
-			},
+		memUsageValue := memStat.Usage()
+		memUsageMetric, err := metriccache.PodMemUsageMetric.GenerateSample(
+			metriccache.MetricPropertiesFunc.Pod(uid), collectTime, float64(memUsageValue))
+		if err != nil {
+			klog.V(4).Infof("failed to generate pod mem metrics for pod , err %v", podKey, err)
```
Comment (saintube) [inline pkg/koordlet/metricsadvisor/collectors/podresource/pod_resource_collector.go:157]: `for pod , err` => `for pod %s, err`

File: pkg/koordlet/resmanager/metrics_query.go [hunk 10]
```diff
@@ -52,59 +51,68 @@ func (r *resmanager) collectorNodeMetrics(start, end time.Time, queryMeta metric
...
-	podsMetrics := make([]*metriccache.PodResourceMetric, 0, len(podsMeta))
+	podsMetrics := make(map[string]float64)
 	for _, podMeta := range podsMeta {
-		podQueryResult := r.collectPodMetric(podMeta, queryParam)
-		podMetric := podQueryResult.Metric
-		if podMetric != nil {
-			podsMetrics = append(podsMetrics, podMetric)
+		queryMeta, err := metricResource.BuildQueryMeta(metriccache.MetricPropertiesFunc.Pod(string(podMeta.Pod.UID)))
+		if err != nil {
+			klog.Warningf("build pod %s/%s query meta failed, error: %v", podMeta.Pod.Name, podMeta.Pod.Namespace, err)
```
Comment (saintube) [inline pkg/koordlet/resmanager/metrics_query.go:65]: the namespace should be ahead of the name. same below

File: pkg/koordlet/resmanager/metrics_query.go [hunk 11]
```diff
@@ -52,59 +51,68 @@ func (r *resmanager) collectorNodeMetrics(start, end time.Time, queryMeta metric
...
+			continue
 		}
+		podQueryResult, err := r.collectPodMetric(queryMeta, *queryParam.Start, *queryParam.End)
+		if err != nil {
+			klog.Warningf("query pod %s/%s  cpu metric failed, error: %v", podMeta.Pod.Name, podMeta.Pod.Namespace, err)
```
Comment (saintube) [inline pkg/koordlet/resmanager/metrics_query.go:70]: remember to format the log message

**Review by jasonliu747 — APPROVED** [review]

**Review by jasonliu747 — APPROVED** [review]

**Review by saintube — COMMENTED** [review]
/lgtm

**Pull request merged**